Update EIP-663: Move to Draft (#7819)
ethereum/EIPs: EIPS/eip-663.md

---
eip: 663
title: Additional stack manipulation instructions
description: Introduce additional stack manipulation instructions which take an immediate arguments for the depth
author: Alex Beregszaszi (@axic), Charles Cooper (@charles-cooper), Danno Ferrin (@shemnon)
discussions-to: https://ethereum-magicians.org/t/eip-663-unlimited-swap-and-dup-instructions/3346

> status: Draft

Reviewer: We hope to get EIP-3540 out of Stagnant soon https://github.com/ethereum/EIPs/pull/8217
then this can be left in Review

Reviewer 2: @charles-cooper :point_up: this can be undone now, 3540 is out of Stagnant.

---
eip: 663

> title: Additional stack manipulation instructions

Reviewer 2: :+1: Andrei's proposal. But also - would just bringing the mnemonics back into the title / description work ("`SWAPN`, `DUPN`, `EXCHANGE`")?

> description: Introduce additional stack manipulation instructions which take an immediate arguments for the depth

Reviewer 3: Same kind of comment for the description.

author: Alex Beregszaszi (@axic), Charles Cooper (@charles-cooper), Danno Ferrin (@shemnon)
discussions-to: https://ethereum-magicians.org/t/eip-663-unlimited-swap-and-dup-instructions/3346

> status: Review

Reviewer 2: 3540 got marked as stagnant (will be fixed soon). Can we in the meantime put this back to `Draft`? Otherwise we can't merge the PR :(

## Abstract

> Currently, `SWAP*` and `DUP*` instructions are limited to a stack depth of 16. Introduce three new instructions, `SWAPN`, `DUPN` and `EXCHANGE` which lift this limitation and allow accessing the stack up to a depth of 32, 32 and 256 items, respectively.

Reviewer 2: `32, 32 and 256` sounds off, shouldn't be `257, 256 and 33`? - however I'm wondering if this isn't too detailed. Maybe can consider making it vague "manipulating stack items at much greater depth"

The number of these arguments (or stack items) can easily exceed 16 and thus will require extra care from a compiler to lay them out in a way that all of them are still accessible.

Lastly, swapping items besides the 1st and Nth items in the stack is very important for compilers implementing stack scheduling algorithms (the analog of register allocation for stack machines), which try to minimize stack traffic given a set of variables and usage analysis.

Introducing `SWAPN`, `DUPN` and `EXCHANGE` will provide an option to compilers to simplify accessing deep stack items.

## Specification

We introduce two new instructions:

 1. `DUPN` (`0xe6`)
 1. `SWAPN` (`0xe7`)

>  2. `EXCHANGE` (`0xe8`)

Reviewer: Conflicts with current propsoal for `DATALOAD`

Reviewer 2: Have we chosen an alternative yet? Need to fix it in the megaspec (https://github.com/ipsilon/eof/issues/43)

Reviewer 2: proposing moving a bunch of new opcodes to the D section here: https://github.com/ipsilon/eof/pull/46

Reviewer 2: DATALOAD is being moved to 0xd0

If the code is legacy bytecode, any of these instructions result in an *exceptional halt*. (*Note: This means no change to behaviour.*)

If the code is valid EOF1, the following rules apply:

 1. The instructions are followed by an 8-bit immediate value, which we call `imm`, and can have a value of 0 to 255.
  1.1 In the case of `DUPN` and `SWAPN`, we introduce the variable `n` which equals to `imm + 1`.
  1.2 In the case of `EXCHANGE`, we introduce the variable `n` which is equal to `imm >> 4 + 1`, and the variable `m` which is equal to `imm & 0x0F + 1` (i.e., the first and second nibbles of `imm`, converted to one-indexing).

 2. Code validation is extended to check that no relative jump instruction (`RJUMP`/`RJUMPI`/`RJUMPV`) targets immmediate values of `DUPN`, `SWAPN` or `EXCHANGE`.

 3. The stack validation algorithm of [EIP-5450](./eip-5450.md) is extended:
  3.1. Before `DUPN` if the current stack height is less than `n`, code is invalid. After `DUPN`, the stack height is incremented.
  3.2. Before `SWAPN` if the current stack height is less than `n + 1`, code is invalid. After `SWAPN`, the stack height is unchanged.

>   3.2. Before `EXCHANGE` if the current stack height is less than `n + m`, code is invalid. After `EXCHANGE`, the stack height is unchanged.

Reviewer 2: `n + m + 1`

 4. Execution rules:
  4.1. `DUPN`: the `n`'th stack item is duplicated at the top of the stack. (*Note: We use 1-based indexing here.*)

>   4.2. `SWAPN`: the `n`'th stack item is swapped with the top of the stack. (*Note: We use 1-based indexing here.*)

Reviewer 2: Shouldn't this be `n+1`th? `SWAPN 0x00` is then equivalent to `SWAP1` and swaps 2nd (1-based) item with top 1st item. 

Note that `DUPN 0x00` is equivalent to `DUP1`.

Similarly `EXCHANGE 0x00` is equivalent to `SWAP1`

Author: thanks, resolved. i must have messed those up when merging different commits i had

>   4.3  `EXCHANGE`: the `n`'th stack item is swapped with the `n + m`'th stack item. (*Note: We use 1-based indexing here.*)

Reviewer: This limits the swapping of `nth` item to only items `[n+1, n+16]`, so it can go 32 items deep but only with max distance 16 between items. E.g. you can't swap 1st with 32d.

Is this acceptable limitation for you?

Author: yes, we might want to change the encoding still.

### Size of immediate argument

For `DUPN` a 16-bit size was considered to accommodate the full stack space of 1024 items, however:

1. that would require an additional restriction/check (`n < 1024`)
2. the 256 depth is a large improvement over the current 16 and the overhead of an extra byte would make it less useful

Similarly for `EXCHANGE`, the proposed scheme allows addressing of 32 items.

### Gas cost

The gas cost for these operations is the same as for existing `DUP*` and `SWAP*` instructions, because they are just implemented as pointer swaps.

### Exchange vs extending swap

As mentioned before, `EXCHANGE` is important to compilers implementing stack scheduling algorithms. Specifically, in the case that a stack item is scheduled to be consumed deeper in the stack (for instance, the 3rd item in the stack needs to be moved into 2nd position in order to be consumed by the next operation), that currently takes three instructions, `SWAP2 SWAP3 SWAP2`. However, in the EVM implementation, the implementation is just a pointer swap, so it could be implemented in a single instruction at no extra runtime cost to the client.

> 

Reviewer 4: This argument here also technically doesn't hold up to scrutiny.
Proper EVM stack scheduling can avoid situations in which you need to exchange two slots neither of which are at the stack top. I.e. if you find yourself in the situation of having to swap the 3rd and 2nd position on stack, then you allocated an inefficient position for the values on stack when you produced them in the first place. I've given a talk at Devconnect Amsterdam about the topic, if you're interested.
Conversely, all values in the EVM are *produced* at the stack top and it thereby cannot be avoided to swap the stack top with a deeper position to create optimal stack layouts in the first place.

So, technically, `SWAPN` is actually the more efficient choice, since the more common and unavoidable operation on stack top requires only two bytes instead of three even for deep access, while conversely patterns like `SWAP2 SWAP3 SWAP2` can generally be avoided by proper stack scheduling at the compiler level.

That being said, having written stack scheduling like that myself, I know that it is not the most trivial thing to do - that's why I've not immediately tried to shut down the idea of `EXCHANGE`: it does allow for simpler more naive stack scheduling albeit at the cost of an increase in code size cost globally compared to optimal scheduling.

But either way, `DUPN` and `EXCHANGE` are not a replacement of `SWAPN`, so it would have to be `EXCHANGE2` as well, no matter what.

Author: > generally be avoided by proper stack scheduling at the compiler level.

the situation can definitely be improved with better algorithms, but optimal stack scheduling, especially across basic blocks and subroutines, is hard. (although i don't have a proof, i suspect it is NP-complete, in the same way that register allocation is). and, even if such a theoretically optimal stack scheduler were available, it is sometimes simply not possible in the general case to schedule things to be in the right order ahead of time, for instance, when side effects are interleaved with stack item production.

> But either way, DUPN and EXCHANGE are not a replacement of SWAPN, so it would have to be EXCHANGE2 as well, no matter what.

having things too deep in the stack is also solvable by better memory spilling algorithms, so i don't think phrasing this choice in absolute terms is very constructive here. but sure, if `EXCHANGE` + `EXCHANGE2` are not considered as taking up too many slots, why not.

Given variable `n`, which equals to `imm + 1`, for `1 <= n <= 256`:

    - `DUPN imm` to fail validation if `stack_height < n`.

>     - `EXCHANGE imm` to fail validation if `stack_height < `.

Reviewer 5: need upper range.  n+m?

Author: added in https://github.com/ethereum/EIPs/pull/7819/commits/e28a856469803c11b0f41f58d1e8bd2dbb1fe71a

Given variable `n`, which equals to `imm + 1`, for `1 <= n <= 256`:

    - `DUPN imm` to fail validation if `stack_height < n`.

>     - `EXCHANGE imm` to fail validation if `stack_height < n + m`.

Reviewer 2: I think my other comment got lost: `SWAPN` cases got lost in merges, this entire section needs a revisit at the end. Also specifically it's `n + m + 1` now for EXCHANGE

>     - `EXCHANGE imm` to fail validation if `stack_height < n + m + 1`.

Reviewer 2: ```suggestion
    - `SWAPN imm` to fail validation if `stack_height < n + 1`.
    - `EXCHANGE imm` to fail validation if `stack_height < n + m + 1`.
```

bring back the cases for SWAPN which got lost

    - `DUPN imm` to increment maximum stack height of a function. Validation fails if maximum stack height exceeds limit of 1023.

>     - `DUPN imm`, `EXCHANGE imm` to fail at run-time if gas available is less than 3.

Reviewer 2: ```suggestion
    - `DUPN imm`, `SWAPN imm`, and `EXCHANGE imm` to fail at run-time if gas available is less than 3.
```

>     - otherwise `DUPN imm` should push the `stack[n]` item to the stack, and `SWAPN imm` should swap `stack[n]` with `stack[n + m]`.

Reviewer 2: ```suggestion
    - otherwise `DUPN imm` should push the `stack[n]` item to the stack, and `SWAPN imm` should swap `stack[n + 1]` with `stack[stack.top()]` and `EXCHANGE imm` should swap `stack[n + 1]` with `stack[n + m + 1]`.
```

Given variable `n`, which equals to `imm + 1`, for `1 <= n <= 256`:

    - `DUPN imm` to fail validation if `stack_height < n`.
    - `SWAPN imm` to fail validation if `stack_height < n + 1`.
    - `EXCHANGE imm` to fail validation if `stack_height < n + m + 1`.
    - `DUPN imm` to increment maximum stack height of a function. Validation fails if maximum stack height exceeds limit of 1023.
    - `DUPN imm`, `SWAPN imm`, and `EXCHANGE imm` to fail at run-time if gas available is less than 3.

>     - otherwise `DUPN imm` should push the `stack[n]` item to the stack, and `SWAPN imm` should swap `stack[n + 1]` with `stack[stack.top()]` and `EXCHANGE imm` should swap `stack[n + 1]` with `stack[n + m + 1]`.

Reviewer: These are off by one I think, because indexing of `n` is 1-based.
```suggestion
    - otherwise `DUPN imm` should push the `stack[n - 1]` item to the stack, and `SWAPN imm` should swap `stack[n]` with `stack[stack.top()]` and `EXCHANGE imm` should swap `stack[n]` with `stack[n + m]`.
```
Maybe better to just omit repeating the spec another time here.

Author: so just remove line 96?

Author: yea, why do the test cases simply repeat the spec -- we should have actual test cases

## Security Considerations

> The authors are not aware of any additional risks introduced here. The EVM stack is fixed at 1024 items and most implementations keep that in memory at all times. This change will increase the stack operations available to compilers.

Reviewer: maybe
```suggestion
The authors are not aware of any additional risks introduced here. The EVM stack is fixed at 1024 items and most implementations keep that in memory at all times. This change will increase the number of easy-to-access stack items.
```
